### ISSUE TITLE
Proposal: enable kerberos auth for docker client

### ISSUE THREAD

**Issue by brianwcook** [docker/distribution#1633]
Currently docker client only accepts user / password combination (basic auth over SSL) in conjunction with a token server.  

I would like to see the ability to use either:
  -a logged in user's kerberos credentials to obtain the oauth token from an authorization service 
  -to spsecify a keytab as a command line option to use for the same.

The flow would follow exactly as in https://docs.docker.com/registry/spec/auth/token/ except that instead of sending basic auth over ssl to the auth service, kerberos would be used to obtain the token, which is then used to retry the original client operation.

**Comment by dmcgowan** [comment]
The linked token specification document describes the protocol used by clients for getting an authorization token and how a registry is expected to verify these tokens. In the latest release of Docker we have expanded this specification to define a long lived token which use between clients and the token server. The mechanisms used to validate credentials and issues tokens is not defined in this specification and the implementations are left up to the token server. If the specification currently makes implementation of a needed authentication method impossible then we are willing to update the protocol specification, but the implementation will always be left up to the token server, additional auth mechanism will not be accepted directly into the registry.

The token server specification is designed to be flexible and allow the token server to support a wide range of authentication methods. Currently Docker Trusted Registry provides many different authentication methods and there is also open source implementations to use such as [cesanta](https://github.com/cesanta/docker_auth). I might suggest requesting support with these token server implementations. If you are looking to support a particular method with a custom token server I might look at our reference [test implementation](https://github.com/docker/distribution/tree/master/contrib/token-server). The amount of code to handle the token protocol is relatively small, most the work would be authorizing with kerberos.

**Comment by brianwcook** [comment]
The issue isn't with the specification of the token server itself, which I agree is open and up to the authentication service itself to decide what forms of auth it can accept.  The issue is the docker client itself, which currently only supports basic auth over ssl to complete a "one step operation" using an auth service.

You could theoretically today setup an auth service, request a token from it and then use that token in the docker client command and successfully complete an operation.  The client does not know how to take a keytab on the command line and when presented with a 401 unauthorized message, present kerberos credentials to the auth server, get a token and retry the operation.  This is what I am after, bringing the functionality of kerberos to parity with basic auth the client.

>  additional auth mechanism will not be accepted directly into the registry.

ack

**Comment by dmcgowan** [comment]
I am not familiar enough with how Kerberos challenges clients to understand what the protocol change would need to be. Even if you could define the protocol change there is no guarantee it would be easy to integrate into the docker client as it also has limited mechanisms for communicating registry challenges back through the docker daemon to the command line client.

> ack

Supporting many different authentications mechanisms in the registry codebase is not manageable, the token specification is designed to decouple this from the registry without being opinionated on how registry authorization should work. We treat tokens and credentials as opaque as we can in this specification (docker client treats tokens as completely opaque, registry never needs to see credentials or long lived tokens). We have found that many users initially dislike this decoupling since everyone wants to deploy as few services as possible, but many of those users end up being very satisfied using a 3rd party token server and having something which is focused entirely on authn/authz.

**Comment by brianwcook** [comment]
sorry - my ACK was ' acknowledged' -  I understand why you are doing that [limiting the auth types the registry accepts] and I think its reasonable.

**Comment by dpal** [comment]
This is a usual situation this is why many projects adopt a pluggable architecture with the basic authentication plugin provided by a project and then others interested in more sophisticated authentication methods like Kerberos, Smart Cards, SAML, etc. add plugins as they see fit. Is this something you would be interested in? If so we can definitely share some experience as we did a lot of similar work in OpenStack recently and collaborated with other projects like MariaDB or PostgressDB.

**Comment by dmcgowan** [comment]
@dpal we are certainly interested in seeing authentication methods be pluggable. Supporting most of the methods you mentioned though would involve plugins on the engine rather than the registry/token-server. If such plugins get added to the engine or some other registry client and an update to the authentication specification is required, we would consider it and make the appropriate updates. While there is no full implementation of the token server in the distribution code base, we consider authenticating with the token server as relevant to the registry authentication specification.

I would suggest proposing these plugins on the [engine](https://github.com/docker/docker/issues) so it can be considered within the scope of the interaction with the user and the interaction with the token server.

**Comment by dpal** [comment]
@dmcgowan not necessarily. The engine can still accept just a token. The question is more about how this token is acquired. If the client can support different authentication methods against a token server the registry engine itself does not need to be changed. So the scope of work is really the token server and the client. Do you agree?

**Comment by dmcgowan** [comment]
I agree that the registry does not need to change, the work to support it would be in the token server. The only impact it would have on this project is possibly adding to the token server specification. The client here would be referring to the docker engine which I agree does need to change. Currently the engine only support username/password as input which can be sent to the token server and username/password or refresh token for storage.

**Comment by dpal** [comment]
I see, so some changes will be needed so how we can make it happen?

**Comment by dmcgowan** [comment]
It would help if I better understand what needs to be sent from the client to the token server to authenticate. If any of the methods describe support username/password authentication you should first add support to a token server and if a more sophisticated method is needed then add a proposal which details what changes to authentication need to happen to support the token server. 

My personal opinion is that more advanced methods of authentication with the token server are best implemented using the Oauth2 authorization code grant type (see https://tools.ietf.org/html/rfc6749#section-4.1.3). Then a client could support any method of authorization without the need for client plugins. This could be used to authenticate directly against an oauth provider (such as google) but could also allow an external tool to handle the authentication portion.

**Comment by dpal** [comment]
If I read https://docs.docker.com/registry/spec/auth/token/ correctly the first step is to submit user name and password to the engine. The engine is the part that currently does the authentication. The authorization service is separate. The authorization and authentication terminology is mixed up a bit.
It is unclear how things actually work. I understand that if you provide user name and password the engine will authenticate you using basic auth and then request a token from the authorization server. 
It is unclear what would happen if the user name and password are not provided. Some clarity will be helpful. 
If I could dream a solution I would suggest that if the request to engine does not have user name and password the engine would redirect client to the authorization (and authentication) server with the request to come back when client gets an Oath2 token (something like this already happening but details escape me a bit). Authorization server would do both authentication and then issue a token. For authentication it can really use PKI, Kerberos, SAML/ECM or even another Oauth2 token from a different scope. For some of those authentication methods client needs to be a bit smarter, for some it does not. I think it should be similar to what keystone client does in OpenStack. This experience can be reused. I would argue that Keystone actually might be a good testing implementation of the authorization server. 
I am not a developer so I can't go further than that but I can connect the engineers to this conversation and let them drive further. But that would make sense if there is a way to not assume user name and password. Is there?

**Comment by tiran** [comment]
Hi

@dpal asked me to join the discussion. I looked into the token based authentication and authorization service a couple of months ago. Please let me rephrase what I know about the token service in order to make sure I got it right.

* The local docker client pass credentials (username, password) to the docker daemon
* Authentication happens between local docker daemon and the token authorization service (AS).
* The AS checks authentication and defines authorization information internally.
* Authorization information are passed from the AS to the docker daemon in form of a scope. The scope contains repository and permission flags (whether push and/or pull is allowed).
* The docker daemon just passes the bearer token to the registry.
* The registry only cares about the bearer token. It doesn't never see the user's credentials.
* A trust relationship between registry and AS is established with PKI.

Does it sound about right?

For Kerberos-based authentication the Docker Registry doesn't need to know about Kerberos at all. Kerberos negotiating could happen in steps 3 and 4 of https://docs.docker.com/registry/spec/auth/token/. In step 4 the AS sends

```
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Negotiate
```
The docker daemon then _somehow_(*) requests a Kerberos ticket, retries the request with the Kerberos auth header to the AS (step 3b) and gets the bearer token (step 4b). X.509 client auth with or without smart cards work similar but without the second request.

(*) I see a couple of issues that need to be addressed first:

1. The docker daemon runs as a different user (usually root) and in a different session than the user. This makes it impossible to get hold of the user's Kerberos TGT when it is stored in the Kernel keyring.
1. By default the local docker socket is only accessible by root. The docker CLI must be run as root, too. This prevents access to the TGT in the Kernel keyring as well. Workaround: docker group, chgrp docker /var/run/docker.socket. This is a potential security issue.
1. I assume the authentication protocol between Docker CLI and Docker daemon is as simple as needed for basic auth.
  * For Kerberos, a full duplex protocol is required. The docker daemon must be able to ask the client to perform operations and send results back.
  * For smart cards it is even more complicated. During the TLS handshake the docker daemon has to talk to the card and perform cryptographic operations on it.

The docker daemon in the middle complicates Kerberos and smartcards. Is there a way to let the docker CLI or user handle the bearer token acquisition?

**Comment by dmcgowan** [comment]
@tiran your understanding is very accurate. The "scope" you mentioned being returned by the AS (token server) is a JWT bearer token which is described here https://docs.docker.com/registry/spec/auth/jwt/. 

The reason I brought up using Oauth2 authorization codes (https://docs.docker.com/registry/spec/auth/jwt/) are because of the problems you mentioned. It is not realistic to have the daemon handle this kind of authentication, even with plugins. There is too many required interactions and each protocol has it differences. Likewise I believe the same is true for doing the authentication directly on the client as the plugin would be responsible for providing protecting keys (such as with TLS), but the client would be handling the network interactions. What authorization code grants provide is a mechanism to completely break up the responsibility such that the client never needs to handle authentication (client in oauth2 terms) and the Kerberos authorization handler (resource owner in oauth2 terms) would never need to handle tokens. Essentially the flow is that the client would get an authorization code from token server, call the authorization handler with the authorization code (this is often down on a web browser but that is not required), user interacts with authorization handler to authenticate and gets authorization code, authorization code is given back to client then normal oauth2 resume (see https://tools.ietf.org/html/rfc6749#section-1.5).  Perhaps I could make a diagram to illustrate this more clearly. This is more how I pictured adding generic authorization support but would need to figure out a user friendly way to exchange authorization codes for authorization grants. With this type of approach a plugin would likely look more like a URI handler.
